src/platform/ESP32/MdnsBasicPacketFilters.h

```diff
@@ -0,0 +1,112 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <inet/EndpointQueueFilter.h>
+#include <inet/IPPacketInfo.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace mdns {
+namespace Minimal {
+
+/**
+ * @brief Basic filter that counts how many pending (not yet dequeued) mDNS packets
+ *        are accumulated, and drops those that would cause crossing of the threshold.
+ *
+ */
+class DropIfTooManyQueuedMdnsPacketFilter : public chip::Inet::EndpointQueueFilter
+{
+    // Repeated here since there are no accessible headers with this
+    static constexpr uint16_t kMdnsPort = 5353u;
+
+public:
+    /**
+     * @brief Initialize the packet filter with a starting limit
+     *
+     * @param maxQueuedPackets - max number of pending-in-queue mDNS packets not yet processed
+     */
+    DropIfTooManyQueuedMdnsPacketFilter(size_t maxQueuedPackets) : mMaxQueuedPackets(maxQueuedPackets) {}
+
+    FilterOutcome FilterBeforeEnqueue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                      const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterAfterDequeue`. We use an atomic for the counter.
+
+        // Non-mDNS is not filtered
+        VerifyOrReturnValue(pktInfo.DestPort == kMdnsPort, FilterOutcome::kAllowPacket);
+
+        if (pktInfo.DestAddress.IsIPv4() && pktInfo.DestPort == kMdnsPort)
+        {
+            if (!mStopLogging)
+            {
+                ChipLogError(Discovery, "Started dropping mDNS from storm.");
+                mStopLogging = true;
+            }
+            return FilterOutcome::kDropPacket;
+        }
+
+        if (mNumQueuedMdnsPackets.load() >= mMaxQueuedPackets)
+        {
+            if (!mStopLogging)
+            {
+                ChipLogError(Discovery, "Started dropping mDNS from storm.");
+                mStopLogging = true;
+            }
+            return FilterOutcome::kDropPacket;
+        }
+
+        ++mNumQueuedMdnsPackets;
+
+        return FilterOutcome::kAllowPacket;
+    }
+
+    FilterOutcome FilterAfterDequeue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterBeforeEnqueue`. We use an atomic for the counter.
+
+        // Non-mDNS is ignored
+        VerifyOrReturnValue(pktInfo.DestPort == kMdnsPort, FilterOutcome::kAllowPacket);
+
+        --mNumQueuedMdnsPackets;
+        if (mNumQueuedMdnsPackets.load() == 0)
+        {
+            if (mStopLogging)
+            {
+                mStopLogging = false;
```

Comment:
If we are really worried about the before/after functions being called at the same time on different threads, this logic around mStopLogging is broken.  By the time we get here we might have already gone back to a state where mNumQueuedMdnsPackets >= mMaxQueuedPackets, no?

Reply:
Added a change which would not break